src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -176,422 +227,317 @@ public void DisconnectFromWorkspace()
-            {
-                using (Logger.LogBlock(FunctionId.Tagger_SyntacticClassification_TagComputer_GetTags, CancellationToken.None))
-                {
-                    if (spans.Count > 0 && _workspace != null)
-                    {
-                        var result = GetTagsWorker(spans);
-                        if (result != null)
-                            return result;
-                    }
+                var oldProject = args.OldSolution.GetProject(args.ProjectId);
+                var newProject = args.NewSolution.GetProject(args.ProjectId);
 
-                    return SpecializedCollections.EmptyEnumerable<ITagSpan<IClassificationTag>>();
-                }
+                // make sure in case of parse config change, we re-colorize whole document. not just edited section.
+                if (Equals(oldProject?.ParseOptions, newProject?.ParseOptions))
+                    return;
+
+                _workQueue.AddWork(_subjectBuffer.CurrentSnapshot);
             }
 
-            private IClassificationService TryGetClassificationService(ITextSnapshot snapshot)
+            private void EnqueueWorkIfThisDocument(DocumentId documentId)
             {
-                var languageServices = _workspace.Services.GetLanguageServices(snapshot.ContentType);
-                if (languageServices == null)
-                    return null;
+                if (_workspace == null)
+                    return;
 
-                return languageServices.GetService<IClassificationService>();
+                var bufferDocumentId = _workspace.GetDocumentIdInCurrentContext(_subjectBuffer.AsTextContainer());
+                if (bufferDocumentId != documentId)
+                    return;
+
+                _workQueue.AddWork(_subjectBuffer.CurrentSnapshot);
             }
 
-            private IEnumerable<ITagSpan<IClassificationTag>> GetTagsWorker(NormalizedSnapshotSpanCollection spans)
+            #endregion
+
+            /// <summary>
+            /// Parses the document in the background and determines what has changed to report to
+            /// the editor.  Calls to <see cref="ProcessChangesAsync"/> are serialized by <see cref="AsyncBatchingWorkQueue{TItem}"/>
+            /// so we don't need to worry about multiple calls to this happening concurrently.
+            /// </summary>
+            private async Task ProcessChangesAsync(ImmutableArray<ITextSnapshot> snapshots, CancellationToken cancellationToken)
             {
-                var classificationService = TryGetClassificationService(spans[0].Snapshot);
-                if (classificationService == null)
-                    return null;
+                // We have potentially heard about several changes to the subject buffer.  However
+                // we only need to process the latest once.
+                Contract.ThrowIfTrue(snapshots.IsDefaultOrEmpty);
+                var currentSnapshot = GetLatest(snapshots);
```

Comment:
This is what allows us to batch up a series of notifications into a single processing step.  Currently, we just ignore all but hte latest snapshot notification.  However, i think this literally coudl be simpler by just having us grab _SubjectBuffer.CurrentSnapshot here on the BG.  This would effectively grab whatever the latest snapshot we truly were at, making it so that the queue is really just a way of saying: hey, wake up and process whatever teh buffer is pointing to.

However, i wasn't certain if that was technically unsafe.  For example, in .net i don't beliee there would be any gurantee that this bg thread would necessarily see changes to ITextBuffer.CurrentSnapshot.  So i went with the safer approach here even though it could potentially be simplified more.

Reply:
Would it help to use something like a CancellationSeries to cancel old work when new work is queued up?

Reply:
AsyncBatchingQueue somewhat does that. The problem with a cancellation series is that it opens us up to race conditions where different events may refer to different text snapshots, and an earlier snapshot may cancel a later snapshot. The current approach uses a little more memory and cpu, but is resilient to that concern.